src/client.rs

```diff
@@ -1366,41 +1360,76 @@ impl ClientProtocol {
     }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 struct BearerChallenge {
-    pub realm: Option<String>,
+    pub realm: Box<str>,
     pub service: Option<String>,
+    #[allow(dead_code)]
     pub scope: Option<String>,
 }
 
-impl Challenge for BearerChallenge {
-    fn challenge_name() -> &'static str {
-        "Bearer"
-    }
+impl TryFrom<&HeaderValue> for BearerChallenge {
+    type Error = String;
 
-    fn from_raw(raw: RawChallenge) -> Option<Self> {
-        match raw {
-            RawChallenge::Token68(_) => None,
-            RawChallenge::Fields(mut map) => Some(BearerChallenge {
-                realm: map.remove("realm"),
-                scope: map.remove("scope"),
-                service: map.remove("service"),
-            }),
+    fn try_from(value: &HeaderValue) -> std::result::Result<Self, Self::Error> {
+        let parser = ChallengeParser::new(
+            value
+                .to_str()
+                .map_err(|e| format!("cannot convert header value to string: {:?}", e))?,
+        );
+        let challenge_opt: Vec<BearerChallenge> = parser
+            .filter_map(|parser_res| {
+                if let Ok(chalenge_ref) = parser_res {
+                    let bearer_challenge = BearerChallenge::try_from(&chalenge_ref);
+                    bearer_challenge.map_or_else(|_e| None, Some)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if challenge_opt.is_empty() {
+            Err("Cannot find Bearer challenge".to_string())
+        } else {
+            Ok(challenge_opt[0].clone())
```

Comment:
Nit: I would use `to_owned()` here instead as it more clearly expresses intent. At first glance I thought you were cloning when we didn't need it. You could also simplify the logic by replacing the `if...then` with:

```rust
challenge_opt.into_iter().nth(0).map(|s| s.to_owned()).ok_or_else(|| "Cannot find Bearer challenge".to_string())
```

Reply:
done